src/target/stm32h7.c

```diff
@@ -0,0 +1,463 @@
+		target_add_commands(t, stm32h7_cmd_list, "STM32H7");
+		break;
+	default:
+		return false;
+	}
+	t->idcode = idcode;
+	return true;
+}
+
+static bool stm32h7_flash_unlock(target *t, uint32_t dest)
+{
+	/* gdb_main calls the erase functions only after a target reset.
+	 * So we run on 64 MHz HSI64 clock and can use WRHIGHFREQ 0 and read
+	 * latency 1 even with small VOS. For bank/mass erase and option write,
+	 * issue explicit reset to the target.*/
+	int offset;
+
+	if (dest < BANK2_START) {
+		offset = 0;
+	} else {
+		offset = 0x100;
+	}
+	DEBUG("ful: dest %08lx, offset %08x\n", dest, offset);
+	target_mem_write32(t, FLASH_ACR, 1);
+	while(target_mem_read32(t, FLASH_SR1 + offset) & FLASH_SR_BSY) {
+		if(target_check_error(t)) {
+			DEBUG("FUL: comm failed\n");
+			return false;
+		}
+	}
+	uint32_t sr = target_mem_read32(t, FLASH_SR1 + offset);
+	if (sr & FLASH_SR_ERROR_MASK) {
+		tc_printf(t, "Error 0x%08lx", sr & FLASH_SR_ERROR_MASK);
+		target_mem_write32(t, FLASH_CCR1 + offset, sr & FLASH_SR_ERROR_MASK);
+		DEBUG("FUL: sr %08lx\n", sr);
+		return false;
+	}
+	if (target_mem_read32(t, FLASH_CR1 + offset) & FLASH_CR_LOCK) {
+		/* Enable FLASH controller access */
+		target_mem_write32(t, FLASH_KEYR1 + offset, KEY1);
+		target_mem_write32(t, FLASH_KEYR1 + offset, KEY2);
+	}
+	return true;
+}
+
+static int stm32h7_flash_erase(struct target_flash *f, target_addr addr,
+							   size_t len)
+{
+	target *t = f->t;
+	int offset, start_sector = (addr - BANK1_START) / FLASH_SECTOR_SIZE;
+	int end_sector = (addr + len - BANK1_START - 1) / FLASH_SECTOR_SIZE;
+	target_mem_write32(t, SRAM_BASE, 0); /* Force rewrite of flash_stub.*/
+	if (addr < BANK2_START) {
+		offset = 0;
+	} else {
+		offset = 0x100;
+		start_sector -= NUM_SECTOR_PER_BANK;
+		end_sector   -= NUM_SECTOR_PER_BANK;
+	}
+	DEBUG("fe: addr %08lx len %08x start %d end %d offset %08x\n",
+		  addr, len, start_sector, end_sector, offset);
+	if (stm32h7_flash_unlock(t, addr) == false) {
+		DEBUG("FE: unlock failed\n");
+		return -1;
+	}
+	for (int i = start_sector; i <= end_sector; i++) {
+		uint32_t cr = FLASH_CR_PSIZE64 | FLASH_CR_SER |
+			(i * FLASH_CR_SNB_1);
+		DEBUG("fe: sector %d\n", i);
+		target_mem_write32(t, FLASH_CR1 + offset, cr);
+		target_mem_write32(t, FLASH_CR1 + offset, cr | FLASH_CR_START);
+		while ((target_mem_read32(t, FLASH_SR1 + offset) & FLASH_SR_BSY)) {
+			if(target_check_error(t)) {
+				DEBUG("fe: comm failed\n");
+				return -1;
+			}
+		}
+		if (target_mem_read32(t, FLASH_SR1 + offset) & FLASH_SR_ERROR_MASK) {
+			DEBUG("fe: error, sr: %08lx\n",
+				  target_mem_read32(t, FLASH_SR1 + offset));
+			return -1;
+		}
+	}
+	return 0;
+}
+
+static int stm32h7_flash_write(struct target_flash *f, target_addr dest,
+                               const void *src, size_t len)
+{
+	int res;
+	target *t = f->t;
+	struct stm32h7_flash *h7 = (struct stm32h7_flash *)f;
+	DEBUG("fw dest %08lx len %04x, ", dest, len);
+	if (target_mem_read32(t, SRAM_BASE) == 0) {
```

Comment:
This seems absurd. Why would we not load the stub here? On first entry here, we have no knowledge of what's expected in the SRAM.